i123643 print java version gradle

reidbaker opened the pull request:
#123643
- Add task to projects evaluated by flutter.gradle that will print the java version.
- Add integration test for the existence of javaVersion and the expected format. 
- Add gradle util to get the gradlew version for a specific platform (gradlew everywhere but windows). 

Why does this code need to exist? 
Figuring out what version of java is used by flutter/gradle is done in a few different ways that are not always aligned. 
See this issue https://github.com/flutter/flutter/issues/122609 ,
this issue https://github.com/flutter/flutter/issues/121501 this feature request https://github.com/flutter/flutter/issues/106416

As examples of why assuming the java version is dangerous. 
This task is code flutter can build upon and is the version gradle is using to build no matter how it is configured. 



## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

reidbaker committed:
Add task to projects evaluated by flutter.gradle that will print the java version.

reidbaker committed:
Add integration test to verify javaVersion output

reidbaker committed:
Merge branch 'master' into i123643-print-java-version-gradle

reidbaker committed:
Merge branch 'master' into i123643-print-java-version-gradle

camsim99 commented on packages/flutter_tools/gradle/flutter.gradle:
```diff
@@ -678,6 +678,17 @@ class FlutterPlugin implements Plugin<Project> {
+    // Format of the output of this task is used by other code in flutter to evaluate java version.
+    private static void addTaskForJavaVersion(Project project) {
+        // Warning: The name of this task is used by other code.
+        project.tasks.register('javaVersion') {
```
How do you feel about calling it (print/show/get)JavaVersion? And is the warning for possible edits to the name?

camsim99 reviewed (approved):
LGTM!

chunhtai commented:
I have a more general question: How will this task be used?

chunhtai commented on packages/flutter_tools/test/integration.shard/android_gradle_java_version.dart:
```diff
@@ -0,0 +1,55 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
The name should end with *_test.dart

reidbaker replied:
good catch. fixed.

reidbaker committed:
Rename test to include _test prefix, update warning to say why the name mattters in gradles javaVersion

reidbaker commented:
> I have a more general question: How will this task be used?

There are a handful of issues relating to identifying or configuring the java sdk. I see this a the source of truth for the version of java gradle is using. In practice I think it will be used to warn on incompatible java versions that other tools have. For example java 17 and gradle 6.7.

gnprice commented:
Running a Gradle task is pretty slow. (Particularly due to #119196; I don't know how slow it would still be after fixing that.) So if you're thinking of a use case within a normal `flutter run` or `flutter build`, I think it'd be best to avoid having to run an extra Gradle task.

What should work to identify the Java version being used by Gradle is:
* The environment that the tool starts `gradlew` with is under the tool's control.
* The behavior of `gradlew` is that it looks at JAVA_HOME, and if it doesn't find an executable file at `"$JAVA_HOME/bin/java"` then it uses the `java` executable found in the PATH.
* So the tool can look at the JAVA_HOME and PATH environment variables it intends to pass to `gradlew`, and identify what `java` executable those point to.
* Then to find the version, one can run that `java` executable with the argument `-version`.

That approach isn't quite ground truth, because in principle the user could have an `android/gradlew` script that does something wild. But there a lot of ways that a user could in principle have made a mess, so it's impossible to be prepared for all of them. And I think this is a pretty unlikely one in practice — I think people generally avoid touching the gradlew scripts except by letting a Gradle upgrade do it.

In principle also a future Gradle version's `gradlew` could change that logic; but I think the behavior of "JAVA_HOME, then fall back to PATH" has been a widespread expectation in the Java world for a long time, and has been very stable in `gradlew`.

I think a task like this one could be useful as ground truth in a debugging situation, though. I'd just want to keep it out of a developer's normal `flutter run` or `flutter build` iteration loop.

reidbaker committed:
wrap println in a doLast and add warning about speed of gradle tasks

reidbaker commented:
> > In practice I think it will be used to warn on incompatible java versions that other tools have. For example java 17 and gradle 6.7.
> 
> In that situation, would this task be able to run? Or would it hit the same unhelpful error message as the main build tasks do?
> 
> My guess (though I don't actually know) would be that the error is happening before Gradle even gets to looking at a specific task, so that this task would hit the same unhelpful error message.

Thanks for these comments. I think you are correct that we will get an error before this code can run which makes this pr less valuable but not so much that I think we should drop this PR. I added a warning about speed and intended use and will not rely on this task for the other work I am doing around java versioning.

reidbaker committed:
Fix android_gradle_version_test to run on windows, add gradlew utility for finding gradlew filename

reidbaker committed:
Merge branch 'master' into i123643-print-java-version-gradle

gnprice reviewed (approved):
LGTM with that plan

reidbaker committed:
Fix formatting

chunhtai commented on packages/flutter_tools/lib/src/android/gradle_utils.dart:
```diff
@@ -290,3 +289,12 @@ void exitWithNoSdkMessage() {
   );
 }
+
+// return gradlew file name based on the platform.
```
```suggestion
// Returns gradlew file name based on the platform.
```

chunhtai reviewed (approved):
LGTM

reidbaker committed:
Fix dart doc formatting for getGradleFileName

reidbaker committed:
Merge branch 'master' into i123643-print-java-version-gradle

reidbaker committed:
Merge branch 'master' into i123643-print-java-version-gradle

MitchellGoodwin committed:
Add adaptive constructor to Radio and RadioListTile (#123816)

Add adaptive constructor to Radio and RadioListTile

gspencergoog committed:
Rename Sample classes (#124080)

Rename Sample classes

vashworth committed:
Better support for wireless devices in IDEs (#123716)

Better support for wireless devices in IDEs

christopherfujino committed:
[flutter_tools] Remove --no-sound-null-safety #4 (#124039)

Re-land of https://github.com/flutter/flutter/pull/123297 without any of
the commits at the end attempting to fix customer-testing.

Fixes https://github.com/flutter/flutter/issues/118810

vashworth committed:
Add flag to filter by device connection interface (#124034)

Add flag to filter by device connection interface

vashworth committed:
Allow --ipv6 flag for flutter attach on iOS devices (#123822)

Allow --ipv6 flag for flutter attach on iOS devices

eyebrowsoffire committed:
Always use the canvaskit path from the web sdk. (#123915)

Always use the canvaskit path from the web sdk.

rydmike committed:
FIX: SegmentedButton segments background cannot be made transparent (#123293)

keyonghan committed:
Move analyzer_benchmark_vm to vm (#124147)

Move analyzer_benchmark_vm to vm

chunhtai committed:
Adds i18n to widget layer (#123620)

Adds i18n to widget layer

TahaTesser committed:
Fix `clipBehavior` for `Drawer` with shape and add `clipBehavior` property. (#124104)

bkonyi committed:
Reland "Serve DevTools when running flutter test (#123607)" (#124058)

Originally reverted due to a misattributed hot reload benchmark
regression.

This reverts commit f5bce716e1fc67abe762f3396627b7e74daa8492.

engine-flutter-autoroll committed:
Roll Flutter Engine from 8edd3c000a39 to 3e5f27d4812e (3 revisions) (#124158)

Roll Flutter Engine from 8edd3c000a39 to 3e5f27d4812e (3 revisions)

engine-flutter-autoroll committed:
Roll Flutter Engine from 3e5f27d4812e to 65515adfb5eb (5 revisions) (#124170)

Roll Flutter Engine from 3e5f27d4812e to 65515adfb5eb (5 revisions)

hellohuanlin committed:
[platform_view]fix iOS platform view's focus node leakage (#124066)

[platform_view]fix iOS platform view's focus node leakage

engine-flutter-autoroll committed:
Roll Flutter Engine from 65515adfb5eb to 4219c72ac8f0 (1 revision) (#124182)

Roll Flutter Engine from 65515adfb5eb to 4219c72ac8f0 (1 revision)

DanTup committed:
[flutter_tools] Include mode in app.start event, and forward app.start to DAP clients (#121239)

[flutter_tools] Include mode in app.start event, and forward app.start to DAP clients

engine-flutter-autoroll committed:
Roll Flutter Engine from 4219c72ac8f0 to db23c3fbe0b0 (4 revisions) (#124200)

https://github.com/flutter/engine/compare/4219c72ac8f0...db23c3fbe0b0

2023-04-05 bdero@google.com [Impeller] Snap glyph positions to screen
space pixels and map UVs correctly (flutter/engine#40912)
2023-04-05 chris@bracken.jp [macOS] Handle termination in
FlutterAppDelegate (flutter/engine#40929)
2023-04-05 skia-flutter-autoroll@skia.org Manual roll Dart SDK from
f97b9d9b2f64 to beff36793081 (1 revision) (flutter/engine#40934)
2023-04-05 chillers@google.com Revert "[web] remove obsolete object
caches; simplify native object management" (flutter/engine#40937)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC jonahwilliams@google.com,rmistry@google.com,zra@google.com on
the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter:
https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

cbracken committed:
[macOS] Remigrate principal class to NSApplication (#124173)

In #122336 we migrated the principal class from NSApplication to
FlutterApplication in the app Info.plist. We removed the need for
FlutterApplication in https://github.com/flutter/engine/pull/40929. This
reverses the migration for anyone who previously upgraded on the Flutter
master branch.

Issue: https://github.com/flutter/flutter/issues/30735

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined
there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my
responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features
we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description
above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is
[test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel
on [Discord].

<!-- Links -->
[Contributor Guide]:
https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]:
https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]:
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]:
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]:
https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

Piinks committed:
Revert "Fix scrolling in the  `Drawer` and `NavigationDrawer` triggers AppBar's scrolledUnderElevation" (#124226)

Revert "Fix scrolling in the  `Drawer` and `NavigationDrawer` triggers AppBar's scrolledUnderElevation"

goderbauer committed:
Add CustomDimensions.commandRunIsTest (#124135)

engine-flutter-autoroll committed:
Roll Flutter Engine from db23c3fbe0b0 to 67467127295d (1 revision) (#124227)

Roll Flutter Engine from db23c3fbe0b0 to 67467127295d (1 revision)

eliasyishak committed:
Add analytics package + setTelemetry method attached (#124015)

The first of many PRs for transitioning to `package:unified_analytics`.
This PR is only focused on disabling and enabling telemetry via the
`flutter config --[no-]analytics` flag

Fixes: https://github.com/flutter/flutter/issues/121617

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined
there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my
responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features
we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description
above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is
[test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel
on [Discord].

<!-- Links -->
[Contributor Guide]:
https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]:
https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]:
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]:
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]:
https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

---------

Co-authored-by: Christopher Fujino <fujino@google.com>
Co-authored-by: Christopher Fujino <christopherfujino@gmail.com>

reidbaker committed:
Add Java-Gradle-AGP validation to flutter analyze (#123916)

https://github.com/flutter/flutter/issues/123917

Doc covering a broad set of issues related to android studio updating.

https://docs.google.com/document/d/1hTXkjbUrBnXgu8NQsth1c3aEqo77rWoEj8CcsQ39wwQ/edit?pli=1#

Specifically this pr:
- Adds new functions to find a projects AGP, Gradle and java versions,
and tests.
- Adds new functions that take versions and parse if the versions are
compatible with each other, and tests.
- Adds validator for `flutter analyze --suggestions` that evaluates the
java/gradle/agp versions and checks if they are compatible, and
integration test.
- Updates the version of gradle used by
dev/integration_tests/flutter_gallery/ to the minimum supported by java
18 so that the integration tests pass (It is unknown why the java
version is 18.9 instead of 11)
- Moves `isWithinVersionRange` to version.dart, and tests.
- Adds FakeAndroidStudio to fakes to be used in multiple tests but does
not remove existing copies.

Metrics will be included as part of the definition of done for this bug
but not as part of this cl. It is already too big.

Known work still left in this pr:
* Understand why analyze integration tests are failing.

Example output if Java and gradle are not compatible:
```
┌───────────────────────────────────────────────────────────────────┐
│ General Info                                                      │
│ [✓] App Name: espresso_example                                    │
│ [✓] Supported Platforms: android                                  │
│ [✓] Is Flutter Package: yes                                       │
│ [✓] Uses Material Design: yes                                     │
│ [✓] Is Plugin: no                                                 │
│ [✗] Java/Gradle/Android Gradle Plugin:                            │
│                                                                   │
│ Incompatible Java/Gradle versions.                                │
│                                                                   │
│ Java Version: 17.0.6, Gradle Version: 7.0.2                       │
│                                                                   │
│ See the link below for more information.                          │
│ https://docs.gradle.org/current/userguide/compatibility.html#java │
│                                                                   │
└───────────────────────────────────────────────────────────────────┘
```
Example output if Gradle and AGP are not compatible
```
┌─────────────────────────────────────────────────────────────────────────────┐
│ General Info                                                                │
│ [✓] App Name: espresso_example                                              │
│ [✓] Supported Platforms: android                                            │
│ [✓] Is Flutter Package: yes                                                 │
│ [✓] Uses Material Design: yes                                               │
│ [✓] Is Plugin: no                                                           │
│ [✗] Java/Gradle/Android Gradle Plugin: Incompatible Gradle/AGP versions.    │
│                                                                             │
│ Gradle Version: 7.0.2, AGP Version: 7.4.2                                   │
│                                                                             │
│ Update gradle to at least "7.5".                                            │
│ See the link below for more information:                                    │
│ https://developer.android.com/studio/releases/gradle-plugin#updating-gradle │
│                                                                             │
│ Incompatible Java/Gradle versions.                                          │
│                                                                             │
│ Java Version: 17.0.6, Gradle Version: 7.0.2                                 │
│                                                                             │
│ See the link below for more information:                                    │
│ https://docs.gradle.org/current/userguide/compatibility.html#java           │
│                                                                             │
└─────────────────────────────────────────────────────────────────────────────┘
```
Example output if Java/Gradle/Agp are not compatible.
```

┌─────────────────────────────────────────────────────────────────────────────┐
│ General Info                                                                │
│ [✓] App Name: espresso_example                                              │
│ [✓] Supported Platforms: android                                            │
│ [✓] Is Flutter Package: yes                                                 │
│ [✓] Uses Material Design: yes                                               │
│ [✓] Is Plugin: no                                                           │
│ [✗] Java/Gradle/Android Gradle Plugin: Incompatible Gradle/AGP versions.    │
│                                                                             │
│ Gradle Version: 7.0.2, AGP Version: 7.4.2                                   │
│                                                                             │
│ Update gradle to at least "7.5".                                            │
│ See the link below for more information:                                    │
│ https://developer.android.com/studio/releases/gradle-plugin#updating-gradle │
│                                                                             │
│ Incompatible Java/Gradle versions.                                          │
│                                                                             │
│ Java Version: 17.0.6, Gradle Version: 7.0.2                                 │
│                                                                             │
│ See the link below for more information:                                    │
│ https://docs.gradle.org/current/userguide/compatibility.html#java           │
│                                                                             │
└─────────────────────────────────────────────────────────────────────────────┘
```

Commit messages
- Add function to gradle_utils.dart that gets the gradle version from
wrapper or system and add a test for each situation
- Add method to get agp version, add method to validate agp against
gradle version, update documentation, add tests for agp validation.
- Update dart doc for validateGradleAndAgp to describe where the info
came from and corner case behavior, create function to validate java and
gradle and hardcode return to false
- Fill out and test java gradle compatibility function in gradle_utils
- Hook up java gradle evaluateion to hasValidJavaGradleAgpVersions with
hardcoded java version
- Add java --version output parsing and tests
- Add getJavaBinary test
- Update comment in android_sdk for mac behavior with java_home -v

- [x] I read the [Contributor Guide] and followed the process outlined
there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my
responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features
we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description
above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is
[test-exempt].
- [ ] All existing and new tests are passing.

reidbaker committed:
Merge remote-tracking branch 'origin/master' into i123643-print-java-version-gradle

christopherfujino commented on packages/flutter_tools/test/integration.shard/android_gradle_java_version_test.dart:
```diff
@@ -0,0 +1,57 @@
+    // Verify that gradlew has a javaVersion task.
+    expect(result.exitCode, 0);
+    // Verify the format is a number on its own line.
+    expect(result.stdout.toString(), matches(RegExp(r'\d\n')));
```
2 questions:

1. should we be matching 1 or more digits?
2. would this never be a `\r\n` on Windows?

reidbaker replied:
1. yes updated. 
2. Maybe I think I will update this to use the end of line sequence instead.

christopherfujino commented on packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart:
```diff
@@ -9,6 +9,7 @@ import 'package:flutter_tools/src/base/logger.dart';
 import 'package:flutter_tools/src/base/platform.dart';
 import 'package:flutter_tools/src/cache.dart';
 import 'package:flutter_tools/src/project.dart';
+import '../../commands.shard/hermetic/build_windows_test.dart' as windows;
```
TIL about this check:

```
▌19:42:44▐ Test imports...
╔═╡ERROR╞═══════════════════════════════════════════════════════════════════════
║ The following file import a test directly. Test utilities should be in their own file.
║ /tmp/flutter sdk/packages/flutter_tools/test/general.shard/android/gradle_utils_test.dart
╚═══════════════════════════════════════════════════════════════════════════════
```

christopherfujino replied:
Failing the analysis shard

reidbaker committed:
Dont depend on another test file

reidbaker committed:
Dont depend on another test file

reidbaker committed:
Merge branch 'master' into i123643-print-java-version-gradle

reidbaker committed:
Use platform path seperator to use .\ on windows

reidbaker committed:
Merge branch 'master' into i123643-print-java-version-gradle
